frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -746,38 +748,37 @@ type dispatchOpt struct {
 	cgroupParent      string
 	llbCaps           *apicaps.CapSet
 	sourceMap         *llb.SourceMap
+	warn              linter.LintWarnFunc
+	argCmdVars        map[string]struct{}
 }
 
 func dispatch(d *dispatchState, cmd command, opt dispatchOpt) error {
+	env, err := d.state.Env(context.TODO())
+	if err != nil {
+		return err
+	}
 	if ex, ok := cmd.Command.(instructions.SupportsSingleWordExpansion); ok {
 		err := ex.Expand(func(word string) (string, error) {
-			env, err := d.state.Env(context.TODO())
-			if err != nil {
-				return "", err
-			}
-			return opt.shlex.ProcessWord(word, env)
+			newword, _, err := opt.shlex.ProcessWord(word, env)
+			return newword, err
 		})
 		if err != nil {
 			return err
 		}
 	}
 	if ex, ok := cmd.Command.(instructions.SupportsSingleWordExpansionRaw); ok {
 		err := ex.ExpandRaw(func(word string) (string, error) {
-			env, err := d.state.Env(context.TODO())
-			if err != nil {
-				return "", err
-			}
-
 			lex := shell.NewLex('\\')
 			lex.SkipProcessQuotes = true
-			return lex.ProcessWord(word, env)
+			newword, _, err := lex.ProcessWord(word, env)
+			return newword, err
 		})
 		if err != nil {
 			return err
 		}
 	}
+	validateCommandVar(cmd.Command, env, &opt)
```

Comment:
unclear why this is not done with previous `lex.ProcessWord` calls and doing replacements on `fmt.Stringer` does not look safe compared to the actual `Expand()` implementation.

Is it because build-args with no value are missing in `Env()`? In that case we can just add them from the `d.buildArgs` as they shouldn't cause any functional difference if their values are empty. (Note that there is also `ProcessWordWithMap` ).

Reply:
That was exactly why, ty!